feat(theme): Update design tokens and defaultTheme

masoudmanson opened the pull request:
## Summary

**Design Tokens**
Github issue: #747 


## Change List

### Storybook

- New story Badges in the preview header (Stable, Beta, Needs Revision, Work in progress)
- New story Badges in the sidebar list (Stable, Beta, Needs Revision, Work in progress)
- New categorization for stories (Bases, Components, Data Viz, Deprecated)
- Added Bases category to Storybook to cover Borders, Colors, Corners, Shadows and Spaces 

### DefaultTheme and Design Tokens

All design tokens have been updated to the latest values based on the Figma Refactor outputs. Token names have been changed to reflect the new semantic structure. DefaultTheme has been created based on new design tokens, and it's ready to implement a multi-theme approach.

### Component Styles

All components have been refactored and are now using the latest design tokens from the palette. All hardcoded spacings, colors, borders, etc., have been replaced with corresponding values from the theme Palette, making the whole design system dynamic and ready to use different theme files.

### Icons

- All SDS icons have been replaced with newly re-exported icons with larger 16 x 16 and 24 x 24 sizes.
- Loading icons have been re worked to show the animation. 
- Icon names are now `PascalCase` instead of `camelCase`
- Some icon names have changed (a full list of changes will be included in the migration documentation).

### Updated Icon sizes for some components

Since the icons are a bit larger by default now, the s size chevron icon in some components has been replaced with xs size for better visuals. 

- Accordion
- ButtonDropdown
- InputDropdown

### Storybook Stories

- Component props are excluded from all stories except for the default story.

### Updated `intent` and `color` props

The accepted values for `intent` / `color` props in SDS components has changed to semantic alternates

- `primary` => `accent`
- `info` => `info`
- `error` => `negative`
- `warning` => `notice`
- `success` => `positive`
- `beta` => `beta`

### Reworked components

Some SDS components have been reworked and bug-fixed.

- `InputCheckbox`, `InputRadio`: Component structure has changed, and this bug (#441) is resolved.
- `InputToggle`
- `InputSearch`: The search icon has been relocated to the left side of the input, and a new Clear icon has been added to the right side.
- `SegmentedControl`
- `Link`: The `Link` component now accepts two new props, `sdsSize` and `fontWeight`
- ‍`Tag`: The `color` prop is no longer required, and the `Tag` component will be rendered as a neutral tag if no colors are provided.
- `CellBasic`: Introducing two additional properties to the `CellBasic` component, enabling the inclusion of elements either to the right or at the bottom of the primary text (#758).

tihuan reviewed (commented):
Thanks so much for the super thorough PR and adding all the comments and PR description too, Masoud ✨ 🥇 🙏 Super helpful!

I think I still have 25% left to review, but thought I'd share the comments I have now, since I think a lot of them can be addressed first

Thanks again!

masoudmanson commented on .storybook/components/SidebarItem/SidebarItem.tsx:
```diff
@@ -0,0 +1,21 @@
+
+const SidebarItem = (props: SidebarItemProps) => {
+  return (
+    <StyledSidebarItem>
+      <div>{props.children}</div>
+      <StatusTag type={props.status}>{props.status}</StatusTag>
+    </StyledSidebarItem>
```
Great catch! I'll make that adjustment. Thanks for the suggestion! 🥳

masoudmanson commented on .storybook/manager-head.html:
```diff
@@ -1,27 +1,30 @@
 <meta property="og:description" content="SDS design system by CZI" />
-<meta property="og:title" content="Vibe Design System" />
```
Haha, I spotted it while sharing a link from the SDS Storybook on Slack! 🤣🤣

masoudmanson commented on .storybook/manager-head.html:
```diff
@@ -1,27 +1,30 @@
 />
-<link rel="preconnect" href="https://fonts.gstatic.com" />
+<link rel="preconnect" href="https://fonts.googleapis.com" />
+<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin />
```
I'm not entirely sure about it. I copied it directly from the Google Fonts embed section. 

You can check it out here for yourself: [Google Fonts - Inter](https://fonts.google.com/specimen/Inter). 
Click on `Get Font` and then on `Get Embed Code.` 

Additionally, this [discussion](https://webmasters.stackexchange.com/questions/89466/when-should-i-use-the-crossorigin-attribute-on-a-preconnect-link) seems to cover the topic.

tihuan replied:
Ahh got it! If we just copied it, then I trust they know what they're doing 😂 Thanks for sharing the links too!

masoudmanson replied:
Haha! well, they sure know what they're doing 😂

masoudmanson commented on packages/components/README.md:
```diff
@@ -199,9 +199,14 @@ To use the default theme in your React application, complete the following:
+/>
 <link
-  href="https://fonts.googleapis.com/css2?family=Open+Sans:ital,wght@0,300;0,400;0,600;0,700;0,800;1,400;1,600;1,700&display=swap"
+  href="https://fonts.googleapis.com/css2?family=IBM+Plex+Mono:wght@400;600&display=swap"
```
It is the selected font family for the Code parts of SDS. 

[Figma Handoff File - Typography](https://www.figma.com/file/CuKxdbQsexhca5hnB4Z9LN/SDS-Figma-Refactor?type=design&node-id=1339-3059&mode=design&t=s2FXyt2yRlxx3jWs-0)

tihuan replied:
ahh I see! Thank you!

masoudmanson commented on packages/components/src/common/customSvgIcon.tsx:
```diff
@@ -7,7 +7,9 @@ import { SvgIcon, SvgIconProps } from "@mui/material";
   return (
     <SvgIcon color="primary" viewBox="2 2 20 20" {...props}>
-      <path d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z" />
+      <path d="M7 19c-1.1 0-2 .9-2 2h14c0-1.1-.9-2-2-2h-4v-2h3c1.1 0 2-.9 2-2h-8c-1.66 0-3-1.34-3-3 0-1.09.59-2.04 1.46-2.56C8.17 9.03 8 8.54 8 8c0-.21.04-.42.09-.62C6.28 8.13 5 9.92 5 12c0 2.76 2.24 5 5 5v2z"></path>
+      <path d="M10.56 5.51C11.91 5.54 13 6.64 13 8c0 .75-.33 1.41-.85 1.87l.59 1.62.94-.34.34.94 1.88-.68-.34-.94.94-.34-2.74-7.53-.94.34-.34-.94-1.88.68.34.94-.94.35z"></path>
+      <circle cx="10.5" cy="8" r="1.5"></circle>
```
The `CustomSvgIcon` is there to showcase the possibility of using a custom SVG icon with SDS components. Originally, it was a house icon, but I've updated it to a more visually appealing science-related icon. 😬

masoudmanson replied:
Old: 
<img width="339" alt="Screenshot 2024-03-04 at 5 37 27 PM" src="https://github.com/chanzuckerberg/sci-components/assets/927990/f0e56788-b43a-48ca-b68a-577cc930891b">

New: 
<img width="334" alt="Screenshot 2024-03-04 at 5 39 41 PM" src="https://github.com/chanzuckerberg/sci-components/assets/927990/9e18be5f-3abd-46ca-b529-2bf6738713ae">

tihuan replied:
OH!! Got it. This is storybook only code, not production code. Maybe worth putting it in a storybook folder?

Maybe not `.storybook` directory, but just `src/common/storybook/*` since it feels like Storybook has full control over creating/destroying `.storybook/`, so probably not safe to save our code there!

masoudmanson replied:
Yes, definitely! That's a good idea. I'll go ahead and move all the Storybook-related files into a dedicated `storybook` folder to keep things neat and organized.

masoudmanson commented on packages/components/src/common/styles-dictionary/design-tokens/borders.json:
```diff
@@ -0,0 +1,44 @@
+  "sds": {
+    "border": {
+      "negative": {
+        "400": { "value": "1px solid {sds.color.primitive.red.400.value}" }
```
The value is coming from [this line](https://github.com/chanzuckerberg/sci-components/blob/6e108696412657e0aed52f2486e4cdcbf00c68cb/packages/components/src/common/styles-dictionary/design-tokens/color.json#L42).

All the design tokens will be merged into one big style-dictionary config because they are all selected [here](https://github.com/chanzuckerberg/sci-components/blob/6e108696412657e0aed52f2486e4cdcbf00c68cb/packages/components/src/common/styles-dictionary/config.js#L48), and we have the option to access them.

tihuan replied:
Ahh I see! So the json file here will still get processed by styles dictionary to produce the final output 💡 Thanks for the education 🚀 !

masoudmanson commented on packages/components/src/core/Accordion/components/AccordionDetails/style.ts:
```diff
@@ -10,8 +10,7 @@ export const StyledAccordionDetails = styled(AccordionDetails)`
     return css`
-      padding: ${spaces?.m}px;
-      padding-top: ${spaces?.xxs}px;
+      padding: ${spaces?.xxs}px ${spaces?.m}px ${spaces?.m}px;
```
Umm, it's a padding shorthand 😬😬

`padding: x y y z;` can be written as `padding: x y z;`

[Source](https://developer.mozilla.org/en-US/docs/Web/CSS/padding#syntax)

tihuan replied:
OH!! Dude I never realized there's the x y z pattern lol TIL thank you!!

masoudmanson commented on packages/components/src/common/SDSBadges.ts:
```diff
@@ -0,0 +1,9 @@
+import { BADGE as addonBadge } from "@geometricpanda/storybook-addon-badges";
```
That makes sense, I will update that. Thanks 🥳✨

masoudmanson replied:
I tried relocating this file into the `.storybook` directory. However, as it is utilized within the `src/core/**/index.stories.tsx` files, I got errors when trying to import it from the `.storybook` directory. 
On the other hand, we already have a `storybookConstants.ts` file in the `src/common` folder with a similar use case. I believe it makes sense to keep this file here as well!

masoudmanson replied:
I renamed it to `storybookBadges.ts` to align with the naming convention used for `storybookConstants.ts`.

tihuan replied:
That works thank you! 

Can you try just `src/common/storybook/*` too? Maybe the build tools don't want us to import from a hidden directory

masoudmanson replied:
Sure, I'll organize storybook files into a dedicated storybook folder.

masoudmanson commented on packages/components/src/core/Accordion/style.ts:
```diff
@@ -73,28 +80,28 @@ const leftPosition = (props: AccordionExtraProps): SerializedStyles => {
-      & .MuiAccordionDetails-root {
-        padding-left: 30px;
+      & .${accordionDetailsClasses.root} {
+        padding-left: 36px;
```
Unfortunately, no, we don't 🥲
But we may be able to compute it using other space tokens! I'll look into it.

masoudmanson replied:
```tsx
// This is to adjust the padding of the AccordionDetails when the togglePosition is left
// The padding-left is the sum of
// the left padding of the AccordionButton = spaces?.m
// the width of the expandIcon = iconSizes?.xs.width
// and the left padding of the AccordionSummary = spaces?.s
& .${accordionDetailsClasses.root} {
  padding-left: ${(spaces?.m ?? 0) +
  (iconSizes?.xs.width ?? 0) +
  (spaces?.s ?? 0)}px;
}
```

tihuan replied:
Wow nice!! So design token driven now 😄 🚀 👏 Ahaha

masoudmanson replied:
Haha :D fully dynamic styles 🥳🎉

masoudmanson commented on packages/components/src/core/Alert/style.ts:
```diff
@@ -34,3 +38,21 @@ export const StyledAlert = styled(Alert)`
   }}
 `;
+
+function SeverityToColor(
```
Right, great catch, will update 👀✨🙇🏻‍♂️

masoudmanson commented on packages/components/src/core/Bases/Borders/style.ts:
```diff
@@ -0,0 +1,42 @@
+
+    const corners = getCorners(props);
+
+    const BorderWidth = 1;
+    const BoxSize = 60;
```
Absolutely, that makes sense.🤩

masoudmanson commented on packages/components/src/core/Bases/Colors/components/Colors/style.ts:
```diff
@@ -0,0 +1,113 @@
+      }
+
+      &:hover {
+        border: solid 1px black;
```
Hmm, sure we can, thanks for spotting it! 🤩🎉

masoudmanson commented on packages/components/src/core/Bases/Colors/components/Colors/index.tsx:
Of course, I'll look into it. 🥳

masoudmanson commented on tsconfig.json:
```diff
@@ -59,12 +59,12 @@
   "references": [
-    { "path": "./packages/components" },
-    { "path": "./packages/data-viz" }
+    // { "path": "./packages/components" },
+    // { "path": "./packages/data-viz" }   
```
Aah, I have to check it. I can't recall why I had to comment these out!

masoudmanson replied:
It looks like we can do away with these references now, as they are causing import errors on components. After extensive testing, I can confirm that removing them doesn't lead to any issues.

tihuan replied:
Sweet! Thanks for investigating and testing this 🙏

tihuan commented on packages/components/src/common/AUTOCOMPLETE_SINGLE_COLUMN_OPTIONS.tsx:
```diff
@@ -34,63 +34,68 @@ export const AUTOCOMPLETE_SINGLE_COLUMN_OPTIONS = [
   },
   {
-    icon: <CustomSdsIcon sdsSize="s" color="gray" />,
+    icon: <CustomSdsIcon sdsSize="s" color="gray" shade={500} />,
```
ooh I see I see. Very convenient! Does the design tell us what shade value to use, or do we just eyeball it?

tihuan commented:
Awesome thanks for updating the PRs and addressing the comments, Masoud! Love this design token driven way so much 🌟 👏 🤩  !

One last request, can you help going through the code to separate storybook only code from the production code like you did in the last commit for some components, so it's consistent too 🙏 ?

I know it's a lot of components to go through, so no rush if you wanna do it in separate PRs too lol

Thanks so much!

masoudmanson commented:
> Awesome thanks for updating the PRs and addressing the comments, Masoud! Love this design token driven way so much 🌟 👏 🤩 !
> 
> One last request, can you help going through the code to separate storybook only code from the production code like you did in the last commit for some components, so it's consistent too 🙏 ?
> 
> I know it's a lot of components to go through, so no rush if you wanna do it in separate PRs too lol
> 
> Thanks so much!

Thanks for reviewing the code, @tihuan. 🎉😍

The latest commit organizes Storybook files and breaks down large Storybook files into smaller and manageable stories. Here is the general structure of each SDS component:

```
Component
 ┣ 📂 __storybook__
 ┃ ┣ 📂 stories
 ┃ ┃ ┣ default.tsx
 ┃ ┃ ┣ livePreview.tsx
 ┃ ┃ ┗ test.tsx
 ┃ ┣ constants.ts
 ┃ ┣ index.stories.tsx
 ┃ ┗ style.ts
 ┣ 📂 __tests__
 ┃ ┣ 📂 __snapshots__
 ┃ ┃ ┗ index.test.tsx.snap
 ┃ ┣ Component.namespace-test.tsx
 ┃ ┗ index.test.tsx
 ┣ 📁 components
 ┣ index.tsx
 ┗ style.tsx
```

tihuan commented on jest.common.config.js:
```diff
@@ -3,6 +3,20 @@ module.exports = {
   preset: "ts-jest",
   setupFilesAfterEnv: ["./jest.setup.ts"],
   testEnvironment: "jsdom",
+  /**
+   * A list of paths to directories that Jest should use to search for files in.
+   * The first two paths are the default paths that Jest uses to search for files.
+   * The third path specifies the namespace-test files that we want to exclude from the test run.
+   * The fourth path specifies the storybook files that we want to exclude from the test run.
+   *
+   * https://jestjs.io/docs/configuration#testmatch-arraystring
```
fabulous comment 🔥 !!

tihuan reviewed (approved):
Wow your new file structure is a piece of art, @masoudmanson ! WE SHOULD FRAME IT 🤩 😆  🖼️ 🥇 👏 

You're one of the best engineers I've had the honor to work with. Thanks so much for all the hard work 🙏 

LGTM TO THE MAX!! 🚀

masoudmanson commented:
> Wow your new file structure is a piece of art, @masoudmanson ! WE SHOULD FRAME IT 🤩 😆 🖼️ 🥇 👏
> 
> You're one of the best engineers I've had the honor to work with. Thanks so much for all the hard work 🙏
> 
> LGTM TO THE MAX!! 🚀

Thank you so much for your kind words! 🤩🥳😍🎉 I appreciate your support. 
So excited to keep pushing forward together! 🚀🎉🥳

clarsen-czi reviewed (changes requested):
Looks so good, @masoudmanson !! Thanks for tackling so many changes! There are only two small things left to fix (font size on the section/column headers in `DropdownMenu` and `XMark` icon size on the `TagFilter`) and then we should be good.

clarsen-czi reviewed (approved):
Looks great @masoudmanson !! Thanks for the attention to detail getting all these changes implemented 💪
